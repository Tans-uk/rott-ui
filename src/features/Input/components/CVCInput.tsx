import React, {type FC} from 'react'

import {StyleSheet, TextInput} from 'react-native'

import type {CVCInputProps} from '../models'
import {InputStyles} from '../styles'
import {InputContainer} from './InputContainer'

export const CVCInput: FC<CVCInputProps> = ({
  fontSize,
  onChangeText,
  theme,
  size,
  label,
  ...props
}) => {
  const handleTextChange = (inputText: string) => {
    let text = inputText.replace(/[^0-9]/g, '')
    text = text.length > 3 ? text.substring(0, 3) : text
    onChangeText!(text)
  }

  return (
    <InputContainer {...props} size={size} theme={theme} label={label}>
      <TextInput
        editable={!props.disabled}
        placeholder='***'
        style={StyleSheet.flatten([InputStyles({fontSize, theme, size}).defaultTextInputStyle])}
        keyboardType='number-pad'
        maxLength={3}
        onChangeText={handleTextChange}
        {...props}
      />
    </InputContainer>
  )
}
