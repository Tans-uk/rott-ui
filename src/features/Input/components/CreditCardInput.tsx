import React, {type FC} from 'react'

import {StyleSheet} from 'react-native'

import type {CreditCardInputProps} from '../models'
import {InputStyles} from '../styles'
import {InputContainer} from './InputContainer'

// Package Imports
import MaskInput from 'react-native-mask-input'

export const CreditCardInput: FC<CreditCardInputProps> = ({
  fontSize,
  onChangeText,
  theme,
  disabled,
  size,
  ...props
}) => {
  const MASK = [
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    '-',
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    '-',
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    '-',
    /\d/,
    /\d/,
    /\d/,
    /\d/,
  ]

  const handleOnChangeText = (text: string) => {
    const replacedText = text.replace(/[^0-9]/g, '')

    onChangeText!(replacedText)
  }

  return (
    <InputContainer {...props} size={size} theme={theme}>
      <MaskInput
        editable={!disabled}
        mask={MASK}
        placeholder='**** **** **** ****'
        keyboardType='number-pad'
        maxLength={19}
        onChangeText={(_masked, unmasked) => handleOnChangeText(unmasked)}
        style={StyleSheet.flatten([InputStyles({fontSize, theme, size}).defaultTextInputStyle])}
        {...props}
      />
    </InputContainer>
  )
}
