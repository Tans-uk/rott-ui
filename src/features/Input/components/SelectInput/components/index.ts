export * from './SelectInput'
export * from './SelectInputModalComponent'
