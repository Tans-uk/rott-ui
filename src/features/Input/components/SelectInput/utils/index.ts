export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
export * from './sortListBySearchPriority'
