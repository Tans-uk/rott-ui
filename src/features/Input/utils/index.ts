export * from './formatByDateMode'
export * from './useInputStyleNormalizer'
