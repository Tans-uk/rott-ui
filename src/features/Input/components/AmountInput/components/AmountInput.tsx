import {useCallback, useContext, useEffect, useRef, useState, type FC} from 'react'

import {StyleSheet, TextInput} from 'react-native'

import {RottUiContext} from '../../../../../contexts'
import {ThemeConfig} from '../../../../../models'
import {Icon} from '../../../../Icon'
import {Item} from '../../../../Item'
import {Label} from '../../../../Label'
import {useInputStyles} from '../../../styles'
import {useInputStyleNormalizer} from '../../../utils'
import type {AmountInputProps} from '../models'
import {AmountInputStyles} from '../styles'

export const AmountInput: FC<AmountInputProps<ThemeConfig>> = ({
  fontSize,
  onChangeText,
  value,
  theme,
  disabled,
  size,
  currencyType = 'TL',
  ...props
}) => {
  const {defaultTextInputStyle} = useInputStyles({fontSize, theme, size})
  const {colors} = useContext(RottUiContext)
  const amountRef = useRef<TextInput>(null)
  const currencyRef = useRef<TextInput>(null)
  const [amount, setAmount] = useState('')
  const [currency, setCurrency] = useState('')

  const resetInternalStates = useCallback(() => {
    setAmount('')
    setCurrency('')
    onChangeText!('0.00')
  }, [])

  const replaceTextWithNumberOrEmpty = (text: string) => {
    const numericOnly = text.replace(/[^0-9]/g, '')

    if (numericOnly === '0') return numericOnly

    return numericOnly.replace(/^0+/, '')
  }

  const amountNormalizer = (text: string) => {
    if (!text) return resetInternalStates()

    const replacedAmount = replaceTextWithNumberOrEmpty(text)
    const reversedText = replacedAmount.split('').reverse().join('') ?? ''
    const dottedText = reversedText?.match(/.{1,3}/g)?.join('.')
    const normalizedText = dottedText?.split('').reverse().join('')

    setAmount(normalizedText ?? '')
  }

  const formatCurrency = (text: string) => {
    return text.replace(/[^0-9]/g, '')
  }

  const handleTextChange = (inputTextAmount: string, inputTextCurrency: string) => {
    if (!inputTextAmount && !inputTextCurrency) return

    let amountFormat = replaceTextWithNumberOrEmpty(inputTextAmount)
    let currencyFormat = formatCurrency(inputTextCurrency)

    amountFormat = amountFormat.length === 0 ? '0' : amountFormat

    currencyFormat = currencyFormat.length === 0 ? '00' : currencyFormat
    currencyFormat = currencyFormat.length === 1 ? `${currencyFormat}0` : currencyFormat

    if (currencyFormat.length > 2) currencyFormat = currencyFormat.substring(0, 2)

    onChangeText!(`${amountFormat}.${currencyFormat}`)
  }

  const placeholderColorNormalizer =
    theme === 'dark'
      ? colors.white
      : (!amount && !currency) ||
          ((amount === '0' || amount === '') && (!currency || currency === '' || currency === '00'))
        ? colors['grey-200']
        : colors['grey-900']

  useEffect(() => {
    const initialAmount = value?.split(',')[0]
    const initialCurrency = value?.split(',')[1]

    setAmount(initialAmount ?? '')
    setCurrency(initialCurrency ?? '')
  }, [])

  useEffect(() => {
    if (value === '' || !value || value === '0.00' || value === '0') resetInternalStates()
  }, [value])

  useEffect(() => {
    handleTextChange(amount, currency)
  }, [amount, currency])

  return (
    <Item relative onTouchStart={() => amountRef.current?.focus()}>
      <Item row {...props}>
        <TextInput
          ref={amountRef}
          nativeID='amount-native-id'
          testID='amount-test-id'
          editable={!disabled}
          placeholder='0'
          maxLength={11}
          value={amount}
          keyboardType='number-pad'
          onChangeText={(text) => amountNormalizer(text)}
          style={StyleSheet.flatten([defaultTextInputStyle, AmountInputStyles().amountInputStyle])}
          placeholderTextColor={placeholderColorNormalizer}
        />

        <Item
          justifyContentFlexEnd
          paddingBottom={useInputStyleNormalizer({size}).bottomElementPadding}>
          <Label
            fontSize={fontSize ?? useInputStyleNormalizer({size}).placeholderSize.toString()}
            fontWeight='bold'
            color={placeholderColorNormalizer as string}>
            ,
          </Label>
        </Item>

        <TextInput
          ref={currencyRef}
          testID='currency-test-id'
          maxLength={2}
          placeholder='00'
          value={currency}
          keyboardType='number-pad'
          onChangeText={(text) => {
            if (text.isEmpty()) amountRef.current?.focus()
            setCurrency(formatCurrency(text).substring(0, 2))
          }}
          onKeyPress={({nativeEvent}) => {
            if (nativeEvent.key === 'Backspace' && currency.isEmpty()) amountRef.current?.focus()
          }}
          onFocus={() => (currency === '00' ? setCurrency('') : undefined)}
          onBlur={() => (currency.isEmpty() || currency === '0' ? setCurrency('00') : undefined)}
          style={StyleSheet.flatten([defaultTextInputStyle, AmountInputStyles().amountInputStyle])}
          placeholderTextColor={placeholderColorNormalizer}
        />
      </Item>

      <Item absolute right={0} bottom={useInputStyleNormalizer({size}).icon.paddingBottom}>
        <Icon
          testID='currency-icon-test-id'
          name={currencyType ?? ''}
          width={useInputStyleNormalizer({size}).icon.width}
          height={useInputStyleNormalizer({size}).icon.height}
          color={colors['grey-200']}
        />
      </Item>
    </Item>
  )
}
