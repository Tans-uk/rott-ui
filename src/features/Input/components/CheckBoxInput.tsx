/* eslint-disable react-native/no-inline-styles */
import {type FC} from 'react'

import {StyleSheet} from 'react-native'

import {themeConfig} from '../../../providers'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Pressable} from '../../Pressable'
import {type CheckBoxInputProps} from '../models'
import {InputStyles} from '../styles'
import {InputStyleNormalizer} from '../utils'
import React from 'react'

export const CheckBoxInput: FC<CheckBoxInputProps> = ({
  description,
  checked,
  fontSize,
  theme,
  size,
  disabled,
  variant,
  fontWeight,
  onCheckChange,
  testID,
  ...props
}) => {
  return (
    <Pressable
      flex={0}
      testID={testID ?? 'checkbox-input-test-id'}
      onPress={() => !disabled && onCheckChange && onCheckChange(!checked)}
      {...props}>
      <Item
        row
        alignItemsCenter
        style={[
          StyleSheet.flatten([InputStyles({theme, size}).defaultTextInputStyle, {height: 'auto'}]),
        ]}
        testID='checkbox-container-test-id'>
        <Item
          borderRadius={8}
          borderColor={checked ? themeConfig.colors.primary : themeConfig.colors['grey-200']}
          borderWidth={2}
          width={24}
          height={24}
          alignItemsCenter
          justifyContentCenter>
          {(checked || disabled) && (
            <Item
              testID='checkbox-checked-test-id'
              borderWidth={2}
              borderColor={
                checked ? themeConfig.colors['neutral-blue-soft'] : themeConfig.colors['grey-200']
              }
              borderRadius={4}
              backgroundColor={
                checked ? themeConfig.colors.primary : themeConfig.colors['grey-200']
              }
              width={checked ? 15 : 22}
              height={checked ? 15 : 22}
            />
          )}
        </Item>

        {description && typeof description === 'string' && (
          <Item flex={1} style={{flexDirection: 'row'}}>
            <Label
              testID='checkbox-default-label-test-id'
              fontSize={fontSize ?? InputStyleNormalizer({size}).placeholderSize}
              fontWeight={fontWeight ?? 'normal'}
              marginLeft={12}
              variant={variant ?? (theme === 'dark' ? 'white' : 'grey-900')}
              style={{flexWrap: 'wrap'}}>
              {description}
            </Label>
          </Item>
        )}

        {description && typeof description !== 'string' && <>{description}</>}
      </Item>
    </Pressable>
  )
}
