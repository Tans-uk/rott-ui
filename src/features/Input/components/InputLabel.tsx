import React, {FC} from 'react'

import {useRottContext} from '../../../hooks'
import {Icon} from '../../Icon'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Pressable} from '../../Pressable'
import {InputLabelProps} from '../models'
import {InputStyleNormalizer} from '../utils'

export const InputLabel: FC<InputLabelProps> = ({
  text,
  disabled,
  icon,
  size,
  theme,
  variant,
  fontFamily,
  description,
  descriptionSize,
  descriptionVariant,
}) => {
  const {language} = useRottContext()
  const getVariant = () => {
    if (variant) return variant
    else if (theme && theme === 'dark') return 'grey-900'
    else if (disabled) return 'grey-200'
    else return 'grey-900'
  }
  const variantCalculation = getVariant()

  return (
    <Item
      row
      paddingHorizontal={InputStyleNormalizer({size: size}).paddingHorizontal}
      alignItemsCenter>
      <Label
        testID='input-label-test-id'
        fontSize={size ? size : 'xs'}
        fontFamily={fontFamily ? fontFamily : 'Markpro-Bold'}
        style={{letterSpacing: 0.5}}
        variant={variantCalculation}>
        {text?.toLocaleUpperCase(language?.name)}
      </Label>

      {description && (
        <Label
          testID='input-label-description-test-id'
          marginLeft={4}
          fontSize={descriptionSize ?? 'xs'}
          fontFamily={fontFamily ?? 'Markpro-Bold'}
          variant={descriptionVariant ?? 'grey-200'}>
          {`(${description?.toLocaleUpperCase(language?.name)})`}
        </Label>
      )}

      {(icon || disabled) && (
        <Pressable
          flex={0}
          alignItemsCenter
          justifyContentCenter
          disabled={disabled}
          marginLeft={4}
          width={icon ? icon?.width && icon.width : 18}
          height={icon ? icon?.height && icon.height : 18}
          onPress={icon ? icon?.onPress : null}>
          <Icon
            testID='input-label-description-icon-test-id'
            name={icon ? icon?.name : 'LOCK'}
            width={icon ? icon?.width : 18}
            height={icon ? icon?.height : 18}
            variant={icon?.variant ?? variantCalculation}
            mode={icon ? icon?.mode : 'stroke'}
            strokeWidth={icon ? icon?.strokeWidth : 2}
            noStroke={icon ? icon?.noStroke : false}
          />
        </Pressable>
      )}
    </Item>
  )
}
