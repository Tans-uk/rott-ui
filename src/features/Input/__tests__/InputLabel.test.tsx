import React from 'react'

import {render} from '../../../__tests__/utils/testUtils'
import {InputLabel} from '../components/InputLabel'

describe('InputLabel', () => {
  const testIds = {
    inputLabelTestId: 'input-label-test-id',
    inputLabelDescriptionTestId: 'input-label-description-test-id',
    inputLabelIconTestId: 'input-label-description-icon-test-id',
  }
  it('ilk render anında snapshot ile eşleşmeli', () => {
    const renderedInputLabel = render(<InputLabel text='Test Label' />)

    expect(renderedInputLabel).toMatchSnapshot()
  })

  describe('InputLabel -> Description', () => {
    it('description verilen değerler ile ekranda görünmeli', () => {
      const {inputLabelDescriptionTestId} = testIds
      const {getByTestId} = render(<InputLabel text='Test Label' description='Test Description' />)
      const inputLabelDescription = getByTestId(inputLabelDescriptionTestId)

      expect(inputLabelDescription).toBeOnTheScreen()
    })

    it('description verilen değerler ile ekranda görünmemeli', () => {
      const {inputLabelDescriptionTestId} = testIds
      const {queryByTestId} = render(<InputLabel text='Test Label' />)
      const inputLabelDescription = queryByTestId(inputLabelDescriptionTestId)

      expect(inputLabelDescription).not.toBeOnTheScreen()
    })
  })

  describe('InputLabel -> Icon', () => {
    it('icon verilen değerler ile ekranda görünmeli', () => {
      const {inputLabelIconTestId} = testIds
      const {getByTestId} = render(<InputLabel text='Test Label' icon={{name: 'INFORMATION'}} />)
      const inputLabelIcon = getByTestId(inputLabelIconTestId)

      expect(inputLabelIcon).toBeOnTheScreen()
    })

    it('icon verilen değerler ile ekranda görünmemeli', () => {
      const {inputLabelIconTestId} = testIds
      const {queryByTestId} = render(<InputLabel text='Test Label' />)
      const inputLabelIcon = queryByTestId(inputLabelIconTestId)

      expect(inputLabelIcon).not.toBeOnTheScreen()
    })
  })
})
