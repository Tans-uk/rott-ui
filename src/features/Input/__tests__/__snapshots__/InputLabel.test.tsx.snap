// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputLabel ilk render anında snapshot ile eşleşmeli 1`] = `
<KeyboardProvider
  enabled={true}
>
  <View
    flex={1}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <View
      alignItemsCenter={true}
      paddingHorizontal={26}
      style={
        {
          "alignItems": "center",
          "flex": undefined,
          "flexDirection": "row",
          "paddingHorizontal": 50,
        }
      }
    >
      <Text
        style={
          {
            "color": "rgba(34, 63, 70, 1)",
            "fontFamily": "Markpro-Bold",
            "fontSize": 10,
            "fontWeight": undefined,
            "letterSpacing": 0.5,
            "textAlign": "auto",
          }
        }
        testID="input-label-test-id"
      >
        TEST LABEL
      </Text>
    </View>
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "alignSelf": "center",
            "elevation": 999999,
            "flex": 0,
            "maxWidth": "100%",
            "position": "absolute",
            "zIndex": 999999,
          },
          {
            "flexDirection": "column-reverse",
            "justifyContent": "flex-start",
            "top": 30,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          {
            "bottom": 0,
          }
        }
      >
        <RCTSafeAreaView />
      </View>
    </View>
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "alignSelf": "center",
            "elevation": 999999,
            "flex": 0,
            "maxWidth": "100%",
            "position": "absolute",
            "zIndex": 999999,
          },
          {
            "bottom": 40,
            "flexDirection": "column",
            "justifyContent": "flex-end",
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          {
            "bottom": 0,
          }
        }
      >
        <RCTSafeAreaView />
      </View>
    </View>
  </View>
</KeyboardProvider>
`;
